### PR TITLE
numbersofinterest: developing numbers of interest algorithm in TDD

### PR DESCRIPTION
**Pull request by aurelienlair**
This pull request is work in progress.
Here is the description of the problem.

Harold always boasted about his prowess with numbers. So one day Reese challenged him to a problem. He gave Harold two numbers X and Y and asked him to find out the Nth number of the series which began with X numbers of Y’s and the following elements are equal to the sum of the last X numbers in the series. Help Harold find the Nth number of the series.

Input:
The first line contains a number T , number of test cases.
The next T lines each contain 3 integers X, Y and N separated by single space.

Output: 
For each test case print the Nth number of the sequence.

Constraints:
1<=T<=10^6
0 < N <= 50
0 < X < 50
0 <= Y< 3

For more details please check:
https://www.hackerearth.com/problem/algorithm/numbers-of-interest-1/

### CODE REVIEW

File: numbersofinterest/NumbersOfInterest.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
+<?php
+
+class NumbersOfInterest
+{
+    public static function calculate($input)
```
Comment (giorgiosironi) [inline numbersofinterest/NumbersOfInterest.php]: would investigate turning input into a data structure (e.g. `class Input`) so that parsing is separated from its usage

File: numbersofinterest/NumbersOfInterest.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    private static function calculateSeriesSequence($series)
+    {
+        list($x, $y, $n) = explode(' ', $series);
+
+        // first part of the series 
+        $currentSeries = array_fill(0, $x, $y);
+
+        for ($cursor=0; $cursor  < ($n-$x); $cursor++) {
+            $sum = 0;
+            for ($z=0; $z<$x; $z++) {
```
Comment (giorgiosironi) [inline numbersofinterest/NumbersOfInterest.php]: would try to extract the second for into a private method, investigate what parameters could be passed

Comment (giorgiosironi) [inline numbersofinterest/NumbersOfInterest.php]: $currentSeries[$z+$cursor] = self::sum(...)

File: numbersofinterest/NumbersOfInterest.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            $currentSeries[$cursor+$x] = $sum;
+        }
+
+        return $currentSeries;
+    }
+}
+
+class SeriesArrayToConsoleLine
+{
+    public static function output(array $series)
```
Comment (giorgiosironi) [inline numbersofinterest/NumbersOfInterest.php]: readable, wonder why only `static` methods

File: numbersofinterest/NumbersOfInterest.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        return implode(
+            PHP_EOL,
+            array_map(
+                function($element)
```
Comment (giorgiosironi) [inline numbersofinterest/NumbersOfInterest.php]: maybe `$series`

File: numbersofinterest/NumbersOfInterestTest.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        // 0 < N <= 50
+        $n = 7;
+
+        // 1 <= T <= 10^6
+        $numberOfSeries = 1;
+
+        $input = "$numberOfSeries" . PHP_EOL . "$x $y $n"; 
+
+        $this->assertEquals(
+            '2, 2, 2, 6, 10, 18, 34',
```
Comment (giorgiosironi) [inline numbersofinterest/NumbersOfInterestTest.php]: separate testing output format from testing the solution is correct

File: numbersofinterest/Algebraist.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
+<?php
+namespace NumbersOfInterest;
+
+require_once __DIR__ . '/SeriesArrayToConsoleLine.php';
+require_once __DIR__ . '/Options.php';
+
+use NumbersOfInterest\Options;
+use NumbersOfInterest\SeriesProperties;
+
+class Algebraist
```
Comment (giorgiosironi) [inline numbersofinterest/Algebraist.php]: of for the abstraction name, quite uncommon but correct

File: numbersofinterest/Algebraist.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
+<?php
+namespace NumbersOfInterest;
+
+require_once __DIR__ . '/SeriesArrayToConsoleLine.php';
+require_once __DIR__ . '/Options.php';
+
+use NumbersOfInterest\Options;
```
Comment (giorgiosironi) [inline numbersofinterest/Algebraist.php]: you don't need this `use` statements, you are already inside the namespace

File: numbersofinterest/Algebraist.php (hunk 6 above)
Comment (giorgiosironi) [inline numbersofinterest/Algebraist.php]: with this abstraction I believe all methods should become non-static

File: numbersofinterest/Algebraist.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    private static function calculateSeriesSequence($seriesOption)
+    {
+        $properties = SeriesProperties::fromArray($seriesOption);
+
+        // first part of the series, fulfill reference number X occurences times 
+        $currentSeries = array_fill(0, $properties->occurencesOfReferenceNumber, $properties->referenceNumber);
+
+        for ($cursor=0; $cursor< ($properties->totalOfNumbers-$properties->occurencesOfReferenceNumber); $cursor++) {
+            // adding sum at position cursor + occurencesOfReferenceNumber
+            $currentSeries[$cursor+$properties->occurencesOfReferenceNumber] = self::sumUntil($properties->occurencesOfReferenceNumber, $currentSeries, $cursor);
```
Comment (giorgiosironi) [inline numbersofinterest/Algebraist.php]: improved, shorter and more abstract method

File: numbersofinterest/Options.php [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+        if ($this->numberOfseries < 1) {
+            throw new \InvalidArgumentException(
+                'Number of series to be performed "' . $this->numberOfSeries . '" not valid (>=1)'
+            );
+        }
+        $this->seriesOptions = $raws; 
+    }
+
+    public function toArray()
```
Comment (giorgiosironi) [inline numbersofinterest/Options.php]: `allSeries()`? `all()`, referring to the options? it's an array of what?

File: numbersofinterest/OptionsTest.php [hunk 10]
```diff
@@ -0,0 +1,56 @@
+<?php
+namespace NumbersOfInterest;
+require_once __DIR__ . '/Options.php';
+use NumbersOfInterest\Options;
+
+class OptionsTest extends \PHPUnit_Framework_TestCase
```
Comment (giorgiosironi) [inline numbersofinterest/OptionsTest.php]: nice to test this in isolation, also with the corner cases such as invalid input

File: numbersofinterest/SeriesPropertiesTest.php [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+     */
+    public function testWhenSeriesPropertiesIsNotValidAnExceptionIsThrown()
+    {
+        new SeriesProperties('');
+    }
+
+    /**
+     * @expectedException InvalidArgumentException
+     */
+    public function testWhenSeriesPropertiesIsNotParsableWithSeparatorAnExceptionIsThrown()
```
Comment (giorgiosironi) [inline numbersofinterest/SeriesPropertiesTest.php]: someone could say that the solution is overengineered due to the higher number of classes; I would try to organize them in folders separating:
- tests from production code
- domain from user interface (input and output management)
